File: kotlin/simple-cipher/src/main/kotlin/SimpleCipher.kt
import kotlin.random.Random

data class Cipher(
    val key: String = generateKey()
) {

    companion object {
        fun generateKey(length: Int = 100): String {
            return buildString {
                repeat(length) {
                    append(('a'.code + Random.nextInt(26)).toChar())
                }
            }
        }
    }

    init {
        require(key.isNotEmpty())
        require(key.all { it in 'a'..'z' })
    }

    fun encode(s: String): String {
        return transform(s, true)
    }

    fun decode(s: String): String {
        return transform(s, false)
    }

    private fun transform(text: String, encode: Boolean): String {
        val result = StringBuilder(text.length)

        for (i in text.indices) {
            val textOffset = text[i] - 'a'
            val keyOffset = key[i % key.length] - 'a'

            val shifted = if (encode) {
                (textOffset + keyOffset) % 26
            } else {
                (textOffset - keyOffset + 26) % 26
            }

            result.append(('a'.code + shifted).toChar())
        }

        return result.toString()
    }
}

File: csharp/simple-cipher/SimpleCipher.cs
using System;
using System.Text;

public class SimpleCipher
{
    public SimpleCipher() => Key = GenerateKey();

    public SimpleCipher(string key) => Key = key;

    public string Key { get; init; }

    public string Encode(string plaintext) => Transform(plaintext, true);

    public string Decode(string ciphertext) => Transform(ciphertext, false);

    private string Transform(string text, bool encode)
    {
        var result = new StringBuilder();

        for (var i = 0; i < text.Length; i++)
        {
            var textOffset = text[i] - 'a';
            var keyOffset = Key[i % Key.Length] - 'a';

            var shifted = encode
                ? (textOffset + keyOffset) % 26
                : (textOffset - keyOffset + 26) % 26;

            result.Append((char)(shifted + 'a'));
        }

        return result.ToString();
    }

    private static string GenerateKey(int length = 100)
    {
        var rnd = new Random();
        var key = new StringBuilder();

        for (int i = 0; i < 100; i++)
        {
            key.Append((char)('a' + rnd.Next(26)));
        }

        return key.ToString();
    }
}